### PR TITLE
feat(video): unify diffusion video encoding across backends

### PR DESCRIPTION
**Pull request by splotnikv**
## Overview:

This unifies the final "raw frames → encoded video"
step across the three diffusion backends behind a single shared encoder in
`dynamo.common`, so codec / container / hardware support is implemented once and
is trivial to extend (new accelerators, new codecs) without touching
backend-specific code.

Design is tracked in a DEP https://github.com/ai-dynamo/enhancements/pull/93.

### Current flow chart

```mermaid
graph LR
    P[Prompt /v1/videos] --> FE[Dynamo Frontend<br/>shared]

    FE --> V[vLLM<br/>AsyncOmni.generate]
    FE --> T[TRT-LLM<br/>DiffusionEngine.generate]
    FE --> S[SGLang<br/>DiffGenerator.generate]

    V --> Ev[export_to_video<br/>ffmpeg / software H.264]
    T --> Et[imageio + ffmpeg<br/>NVENC H.264]
    S --> Es[imageio + ffmpeg<br/>NVENC H.264]

    Ev --> W[upload_to_fs / fs.pipe<br/>shared]
    Et --> W
    Es --> W
    W --> D[(file:// / s3:// …)]
```

The fork in the middle (three encoders) is the only real divergence — and the
target of this work.

## Proposed Architecture

Insert a single shared encoder between the backends and the existing file
writer. Each backend feeds its raw pixels through a **thin conversion layer**
(normalize to a canonical `np.ndarray (T, H, W, 3) uint8`), then calls the
**unified encoder**, which dispatches to the right backend implementation.

```mermaid
graph LR
    V[vLLM frames] --> C[Thin conversion layer<br/>→ numpy T,H,W,3 uint8]
    T[TRT-LLM frames] --> C
    S[SGLang frames] --> C

    C --> U{Unified encoder<br/>dispatch}
    U -->|NVIDIA| N[imageio → ffmpeg<br/>NVENC]
    U -->|new HW / codecs| F[ffmpeg CLI<br/>raw pipe]

    N --> W[upload_to_fs / fs.pipe<br/>unchanged]
    F --> W
    W --> D[(file:// / s3:// …)]
```

## Details:

- Adds to `dynamo/common/utils/video_utils.py`:
  - **Thin conversion layer** (`to_canonical_frames`) that normalizes each
    backend's raw output — TRT-LLM `torch.Tensor (1,T,H,W,C)`, vLLM
    `stage_output.images` list, SGLang `list[PIL | np.ndarray]` — to a canonical
    `np.ndarray (T, H, W, 3) uint8`.
  - **Unified encoder** (`encode_video`) with two paths: imageio → ffmpeg (NVENC),
    preserving existing NVIDIA behavior; and an ffmpeg-CLI raw-pipe path for new
    hardware (Intel XPU / VA-API) and codecs (HEVC).
  - **Env-var controls**: `DYN_VIDEO_CODEC`, `DYN_VIDEO_CONTAINER`,
    `DYN_VIDEO_HW_ACCEL` (`auto`/`nvenc`/`xpu`/`cpu`, auto-detects XPU),
    `DYN_VIDEO_DEVICE` (falls back to legacy `DYNAMO_VAAPI_DEVICE`).
- Wires all three handlers to `encode_video()`:
  - TRT-LLM `video_handler.py` (was `encode_to_video_bytes`).
  - SGLang `video_generation_handler.py` — removes the duplicated inline
    `_frames_to_video` encoder.
  - vLLM-Omni `output_formatter.py` — replaces the `export_to_video` temp-file path;
    XPU is now auto-detected here.
- The storage stage (`upload_to_fs`) is unchanged. Legacy helpers
  (`encode_to_video_bytes`, `encode_to_mp4`, `normalize_video_frames`,
  `frames_to_numpy`) are intentionally kept for now.
- No tests changed. Known impact: `test_trtllm_video_diffusion.py` patches the old
  `encode_to_video_bytes` symbol and will need repatching to `encode_video`
  (documented in the DEP's Testing Impact section). Test migration is deferred.

## Where should the reviewer start?

- `components/src/dynamo/common/utils/video_utils.py` — the new
  `to_canonical_frames` and `encode_video` (dispatch + two encode paths) are the
  core of the change.
- The three handler call sites:
  - `components/src/dynamo/trtllm/request_handlers/diffusion/video_handler.py`
  - `components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py`
  - `components/src/dynamo/vllm/omni/output_formatter.py`

## Related Issues

**🚫 This PR is NOT linked to an issue**:
- [x] Confirmed — no related issue

## Summary by CodeRabbit

* **New Features**
  * Added unified video encoding with automatic or explicit CPU, NVENC, and XPU acceleration support.
  * Added consistent conversion for video frames from images, tensors, and arrays into RGB format.
  * Improved video generation across SGLang, TensorRT-LLM, and vLLM-Omni integrations.
  * Added support for configurable containers, codecs, frame rates, and encoding devices.

* **Bug Fixes**
  * Improved validation of video dimensions, channels, and data types.
  * Added clearer handling when required video encoding tools are unavailable.

* **Tests**
  * Expanded coverage for encoding paths, frame conversion, hardware selection, and video round trips.

### CODE REVIEW

**Comment by ayushag-nv** [comment]
@splotnikv Thanks for flagging this and putting up a great design. I am aligned with your design. Please go ahead for the rest of the implementation. I would love to see some real validations and tests in all three frameworks , so that we are making sure quality is not compromised before and after.

File: components/src/dynamo/common/utils/video_utils.py [hunk 1]
```diff
@@ -242,3 +245,275 @@ def encode_to_video_bytes(
...
+    """Map (container, logical codec, hw path) to an ffmpeg encoder name.
+
+    A codec value that is already an ffmpeg encoder name passes through.
+    """
+    codec = codec or _DEFAULT_CODEC.get(container, "h264")
+    table = _FFMPEG_ENCODERS.get(hw_accel, _FFMPEG_ENCODERS["nvenc"])
+    return table.get(codec, codec)
+
+
+def to_canonical_frames(frames) -> np.ndarray:
```
Comment (ayushag-nv) [inline components/src/dynamo/common/utils/video_utils.py]: This function needs to be extremely robust and well tested

**Comment by rmccorm4** [comment]
Sharing a related PR here just as a heads up: https://github.com/ai-dynamo/dynamo/pull/11492

File: components/src/dynamo/trtllm/request_handlers/diffusion/video_convert.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+def to_canonical(video) -> np.ndarray:
+    """Convert ``VisualGenOutput.video`` to canonical ``(T, H, W, 3) uint8``.
+
+    ``VisualGenOutput.video`` is a ``torch.Tensor`` of shape ``(1, T, H, W, C)``
+    (uint8) since TRT-LLM rc9. The batch dim is squeezed and the tensor moved to
+    host memory before canonicalizing.
+    """
+    assert (
+        video.ndim == 5 and video.shape[0] == 1
+    ), f"Expected video shape (1, T, H, W, C), got {tuple(video.shape)}"
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/trtllm/request_handlers/diffusion/video_convert.py]: fixed in next commit

File: components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py [hunk 3]
```diff
@@ -254,70 +255,11 @@ async def _generate_video(
         if not frames:
             raise RuntimeError("DiffGenerator returned no frames")
 
-        # Convert frames to video bytes
-        video_bytes = await self._frames_to_video(frames, fps)
+        # Convert to canonical frames, then encode via the shared unified encoder.
+        canonical = to_canonical(frames)
+        video_bytes = await asyncio.to_thread(encode_video, canonical, fps)
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py]: Good catch, there is no container / codec compatibility test, added in next commit.

File: components/src/dynamo/common/utils/video_utils.py [hunk 4]
```diff
@@ -242,3 +245,297 @@ def encode_to_video_bytes(
...
+            "-s",
+            f"{width}x{height}",
+            "-r",
+            str(fps),
+            "-i",
+            "-",
+        ]
+        if hw_accel == "xpu":
+            cmd += ["-vf", "format=nv12,hwupload"]
+        cmd += ["-c:v", encoder]
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/common/utils/video_utils.py:489]: done

File: components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py [hunk 5]
```diff
@@ -254,70 +255,11 @@ async def _generate_video(
...
         return video_bytes
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py:263]: It is intentional, added comment to clarify behavior.

File: components/src/dynamo/common/tests/test_video_utils.py [hunk 6]
```diff
@@ -154,3 +162,276 @@ def test_v2_api_fallback_writes_all_frames(self):
...
+        except Exception as e:  # decoder/plugin not available in this env
+            pytest.skip(f"video decode unavailable: {e}")
+
+        assert decoded.shape[0] == num_frames
+        assert tuple(decoded.shape[1:3]) == (height, width)
+
+        # Fidelity: a smooth clip through H.264 should reconstruct well.
+        # 35 dB is a deliberately lenient floor (real content typically >40 dB).
+        psnr = _psnr(frames, decoded)
+        assert psnr >= 35.0, f"round-trip PSNR too low: {psnr:.1f} dB"
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/common/tests/test_video_utils.py:490]: added

File: components/src/dynamo/common/utils/video_utils.py [hunk 7]
```diff
@@ -242,3 +245,297 @@ def encode_to_video_bytes(
...
+
+
+def _running_on_xpu() -> bool:
+    """Return True when torch reports an available XPU backend."""
+    try:
+        import torch
+
+        return bool(hasattr(torch, "xpu") and torch.xpu.is_available())
+    except Exception:
+        return False
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/common/utils/video_utils.py:322]: "Re-raise after logging" is wrong for a capability probe. The guideline is about not hiding real errors, but _running_on_xpu is a best-effort "is XPU there?" check whose entire contract is return a bool, never throw. Re-raising would defeat its purpose — the caller would crash on auto resolution instead of falling back. So the guideline's re-raise clause simply doesn't apply to this pattern.

The failure is not silent-wrong. If the probe wrongly returns False on a real XPU box, auto → nvenc, and the NVENC encode then fails loudly downstream (imageio/ffmpeg: h264_nvenc unavailable). So a swallowed probe error degrades to a meaningful error soon after, exactly as you said — not corrupted output.

File: components/src/dynamo/common/utils/video_utils.py [hunk 8]
```diff
@@ -242,3 +245,297 @@ def encode_to_video_bytes(
...
+    else:
+        writer = iio.get_writer(  # type: ignore[attr-defined]
+            buffer, format="FFMPEG", mode="I", fps=fps, codec=encoder
+        )
+        try:
+            for frame in frames:
+                writer.append_data(frame)
+        finally:
+            writer.close()
+    return buffer.getvalue()
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/common/utils/video_utils.py:446]: fixed

File: components/src/dynamo/common/utils/video_utils.py [hunk 9]
```diff
@@ -242,3 +245,297 @@ def encode_to_video_bytes(
...
+                f"ffmpeg {encoder} encode failed (exit {proc.returncode}): {stderr}"
+            )
+        with open(output_path, "rb") as fh:
+            return fh.read()
+    finally:
+        try:
+            os.unlink(output_path)
+        except OSError:
+            pass
+
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/common/utils/video_utils.py:526]: I've considered adding a timeout to the subprocess.run ffmpeg call but decided against it for now. ffmpeg is a stable, well-tested binary — a genuine hang would signal a system/driver-level fault that a per-call timeout can't meaningfully recover from, and we already bound worst-case duration with the request-level timeout that covers the whole handler. A fixed encode timeout is also hard to pick correctly since encode time scales with frame count and resolution, so a conservative value risks killing legitimate large encodes. If thread-pool exhaustion from stuck encodes shows up in practice, we'll revisit with a configurable ceiling rather than a guessed constant.

File: components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/sglang/request_handlers/video_generation/video_generation_handler.py:263]: fixed

File: components/src/dynamo/trtllm/request_handlers/diffusion/video_convert.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/trtllm/request_handlers/diffusion/video_convert.py]: fixed

File: components/src/dynamo/trtllm/request_handlers/diffusion/video_handler.py [hunk 10]
```diff
@@ -250,22 +251,19 @@ async def generate(
...
-                ), f"Expected video shape (1, T, H, W, C), got {video.shape}"
-                frames_np = video[0].cpu().numpy()
+                # VisualGenOutput.video is (1, T, H, W, C) uint8 since TRT-LLM
+                # rc9; the backend converter squeezes the batch dim and moves
+                # the frames to host memory as canonical (T, H, W, 3) uint8.
+                frames_np = to_canonical(output.video)
                 logger.info(
                     f"Request {request_id}: encoding video output "
                     f"(shape={frames_np.shape}) to MP4 at {fps} fps"
                 )
```
[the remark this answers is not in the document]
Comment (splotnikv) [reply components/src/dynamo/trtllm/request_handlers/diffusion/video_handler.py:263]: fixed

File: components/src/dynamo/common/utils/video_utils.py (hunk 1 above)
Comment (splotnikv) [reply components/src/dynamo/common/utils/video_utils.py]: Done. This function has been refactored, split to common part (backend agnostic) and backend specific format conversion. 4 groups of tests have been added - for common part and for each backend. DEP has been updated accordingly.

**Comment by splotnikv** [comment]
Hi @ayushag-nv , I've addressed your feedback, please have another look.
